File: src/com/mobiletheatertech/plot/SetPlatform.java
package com.mobiletheatertech.plot;

import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import java.util.ArrayList;

/**
 * Created by dhs on 1/21/15.
 */
public class SetPlatform extends MinderDom implements Legendable {

    ArrayList<Shape> Polygons = new ArrayList<>();

    private Double x = null;
    private Double y = null;
    private Double orientation = null;

    public SetPlatform( Element element )
            throws AttributeMissingException, DataException, InvalidXMLException
    {
        super( element );

        x = getDoubleAttribute( "x" );
        y = getDoubleAttribute( "y" );
        orientation = getOptionalDoubleAttributeOrZero( "orientation" );

        ArrayList<Element> polygonList = subElements( element, "shape" );
        for( Element polygonElement : polygonList ) {
            String polygonString = polygonElement.getAttribute("polygon");
            if (null != polygonString && ! "".equals( polygonString )) {
                Shape polygon = new Shape( polygonString );
                Polygons.add( polygon );
            }
        }
    }

    ArrayList<Element> subElements( Element element, String tag ) {
        ArrayList<Element> resultList = new ArrayList<>();

        NodeList displays = element.getElementsByTagName( tag );
        int length = displays.getLength();
        for (int index = 0; index < length; index++) {
            Node node = displays.item( index );

            if (null != node) {
                if (node.getNodeType() == Node.ELEMENT_NODE) {
                    Element subElement = (Element) node;
                    resultList.add(subElement);
                }
            }
        }

        return resultList;
    }

    @Override
    public void verify() throws DataException, FeatureException, InvalidXMLException, LocationException, MountingException, ReferenceException {

    }

    @Override
    public void dom(Draw draw, View mode)
            throws InvalidXMLException, MountingException, ReferenceException
    {
        if( ! Proscenium.Active() ) { return; }


        SvgElement group = draw.group( draw, "" );
        for ( Shape thing : Polygons ) {

            group.path( draw, thing.toSvgPath( x, y ), "black" );
        }
    }

    @Override
    public void countReset() {

    }

    @Override
    public PagePoint domLegendItem(Draw draw, PagePoint start) {
        return null;
    }
}

File: src/com/mobiletheatertech/plot/SetPlatform.scala
package com.mobiletheatertech.plot

import org.w3c.dom.Element
import java.util.ArrayList

import scala.collection.JavaConversions._

/**
  * Created by DHS on 8/11/16.
  */

class SetPlatform (val element: Element, parent: SetPiece ) extends MinderDom(element)
  with Populate
  with Legendable
{
  def this( element: Element ) {
    this( element, null )
  }

  var x: Double = 0.0
  var y: Double = 0.0
  if( null != parent ) {
    x = parent.origin.x
    y = parent.origin.y()
  }

  x += getDoubleAttribute("x")
  y += getDoubleAttribute("y")

  val orientation = getOptionalDoubleAttributeOrZero("orientation")

  val shapes: ArrayList[ Shape ] = new ArrayList[ Shape ]

  tagCallback( Shape.Tag, processShape )
  populate( element )

  if ( shapes.size() < 1 )
    throw new InvalidXMLException(
      "SetPlatform at (" + x + ", " + y + ") has no Shape." )

  println( this.toString )

  def processShape( element: Element ): Unit = {
    shapes.add( new Shape( element ) )
  }

  def verify()
  {
  }

  def dom(draw: Draw, mode: View)
  {
    if( mode == View.SECTION ) return

    val group: SvgElement = draw.group(draw, SetPlatform.Tag )
    group.element.setAttribute( "stroke", SetPlatform.Color )

    for ( thing <- shapes) {
      val element = thing.toSvg( group, draw, x, y )
      element.attribute( "transform", "rotate(" + orientation + "," + x + "," + y + ")" )
      element.attribute( "stroke-width", "1" )
      element.attribute("fill", SetPlatform.Color )
      element.attribute("fill-opacity", "0.1")
    }
  }

  def legendCountReset()
  {
  }

  def domLegendItem(draw: Draw, start: PagePoint): PagePoint =
    null

  override def toString: String = {
    var result: String = "SetPlatform at (" + x + ", " + y +
      ") has " + shapes.size() + " Shapes."
    if (null != parent) {
      result += "\n  Parent: " + parent.toString
    }

    result
  }

}

object SetPlatform {
  val Tag = "set-platform"
  val Color = "green"
}
